document-aligner/docalign.cpp

```diff
@@ -202,17 +206,30 @@ int main(int argc, char *argv[])
 				for (Line const &line : *line_batch) {
 					Document document;
 					ReadDocument(line.str, document, ngram_size);
-					for (auto const &entry : document.vocab)
-						local_df[entry.first] += 1; // Count once every document
+					for (auto const &vocab_entry : document.vocab) {
+						df_entry.ngram = vocab_entry.first;
+						NGramFrequencyMap::MutableIterator df_it;
+						if (local_df.FindOrInsert(df_entry, df_it)) {
+							df_it->count += 1; // Count once every document (as opposed to ngram->)
+						}
+					}
 				}
 			}
 
 			// Merge the local DF into the global one. Multiply by df_sample_rate
 			// to compensate for reading only nth part of the whole collection.
 			{
 				unique_lock<mutex> lock(df_mutex);
-				for (auto const &entry : local_df)
-					df[entry.first] += entry.second * df_sample_rate;
+				for (NGramFrequencyMap::ConstIterator local_it = local_df.RawBegin(); local_it != local_df.RawEnd(); ++local_it) {
+					if (local_it->GetKey() != NGram()) {
+						NGramFrequencyMap::MutableIterator df_it;
+						if (df.FindOrInsert(*local_it, df_it)) {
+							df_it->count += local_it->count * df_sample_rate;
+						} else {
+							df_it->count *= df_sample_rate;
+						}
+					}
+				}
```

Comment:
Why is this so slow? Are there too many empty entries? Or too many collisions which makes `FindOrInsert()` slow?

Reply:
Got stats for one `local_df`:
```
Offsets:
  skip 0: 14702249 times
  skip 1: 3326865 times
  skip 2: 1398420 times
  skip 3: 745469 times
  skip 4: 448572 times
  skip 5: 292315 times
  skip 6: 200919 times
  skip 7: 143398 times
  skip 8: 104909 times
  skip 9: 78907 times
  skip 10: 60797 times
  skip 11: 46818 times
  skip 12: 36798 times
  skip 13: 29191 times
  skip 14: 23707 times
  skip 15: 19090 times
  skip 16: 15525 times
  skip 17: 12820 times
  skip 18: 10665 times
  skip 19: 8750 times
  skip 20: 7344 times
  skip 21: 6123 times
  skip 22: 5150 times
  skip 23: 4213 times
  skip 24: 3636 times
  skip 25: 3057 times
  skip 26: 2649 times
  skip 27: 2211 times
  skip 28: 1952 times
  skip 29: 1647 times
  skip 30: 1420 times
  skip 31: 1156 times
  skip 32: 1107 times
  skip 33: 937 times
  skip 34: 825 times
  skip 35: 677 times
  skip 36: 550 times
  skip 37: 498 times
  skip 38: 443 times
  skip 39: 406 times
  skip 40: 358 times
  skip 41: 309 times
  skip 42: 251 times
  skip 43: 249 times
  skip 44: 214 times
  skip 45: 166 times
  skip 46: 172 times
  skip 47: 154 times
  skip 48: 131 times
  skip 49: 110 times
  skip 50: 88 times
  skip 51: 92 times
  skip 52: 83 times
  skip 53: 56 times
  skip 54: 59 times
  skip 55: 54 times
  skip 56: 38 times
  skip 57: 52 times
  skip 58: 21 times
  skip 59: 23 times
  skip 60: 24 times
  skip 61: 24 times
  skip 62: 19 times
  skip 63: 17 times
  skip 64: 22 times
  skip 65: 26 times
  skip 66: 13 times
  skip 67: 17 times
  skip 68: 6 times
  skip 69: 9 times
  skip 70: 7 times
  skip 71: 7 times
  skip 72: 7 times
  skip 73: 4 times
  skip 74: 7 times
  skip 75: 6 times
  skip 77: 4 times
  skip 78: 2 times
  skip 79: 3 times
  skip 80: 5 times
  skip 81: 3 times
  skip 82: 2 times
  skip 83: 2 times
  skip 84: 2 times
  skip 85: 1 times
  skip 86: 1 times
  skip 87: 2 times
  skip 88: 1 times
  skip 89: 1 times
  skip 91: 1 times
  skip 93: 3 times
  skip 96: 1 times
  skip 98: 1 times
  skip 100: 2 times
  skip 101: 1 times
  skip 114: 1 times
  skip 18446744073709551615: 11799313 times (this is the invalid key option)
  ```

Reply:
I'm running it now for `df` afterwards to see if there's something fishy going on after trying to merge `local_df` into it. But it takes a really long time for some reason. Getting those stats doesn't, so it isn't just the iteration that's slow… it seems to be either the insertion or some weird memory interaction between the two?